source/common/network/utility.cc

```diff
@@ -547,6 +548,59 @@ Api::IoCallUint64Result Utility::readFromSocket(IoHandle& handle,
                                                 UdpPacketProcessor& udp_packet_processor,
                                                 MonotonicTime receive_time,
                                                 uint32_t* packets_dropped) {
+
+  if (handle.supportsUdpGro()) {
+    Buffer::InstancePtr buffer = std::make_unique<Buffer::OwnedImpl>();
+    Buffer::RawSlice slice;
+
+    const uint64_t maxPacketSizeWithGro = 65535;
+    const uint64_t num_slices = buffer->reserve(maxPacketSizeWithGro, &slice, 1);
+    ASSERT(num_slices == 1u);
+
+    IoHandle::RecvMsgOutput output(1, packets_dropped);
+
+    Api::IoCallUint64Result result =
+        handle.recvmsg(&slice, num_slices, local_address.ip()->port(), output);
+
+    if (!result.ok()) {
+      return result;
+    }
+
+    ENVOY_LOG_MISC(trace, "recvmsg bytes {}", result.rc_);
+
+    // Get gso_size
+    if (output.msg_[0].gso_size_ == 0u) {
+      // Skip gso segmentation and proceed as a single payload.
+      ENVOY_LOG_MISC(warn, "Unspecified or zero gso_size received.");
+      passPayloadToProcessor(
+          result.rc_, slice, std::move(buffer), std::move(output.msg_[0].peer_address_),
+          std::move(output.msg_[0].local_address_), udp_packet_processor, receive_time);
+      return result;
+    }
+
+    slice.len_ = std::min(slice.len_, static_cast<size_t>(result.rc_));
```

Comment:
I guess result.rc_ must be <= slice.len_?

If that's true, we can just set slice.len_ to result.rc_.

Reply:
Yes this makes sense here! Changed this accordingly.